### PR TITLE
refactor(migrations): extend abstract migration

### PR DESCRIPTION
**Pull request by pret-a-porter**
## Changes:

1. Extend AbstractMigration
2. Align MigrationsService and existing migrations by new changes

## Context:

1. https://github.com/renovatebot/renovate/issues/11459
2. Part of https://github.com/renovatebot/renovate/pull/12957 

## Documentation

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/migration.ts [hunk 1]
```diff
@@ -33,7 +33,7 @@ export function migrateConfig(
         optionTypes[option.name] = option.type;
       });
     }
-    const newConfig = MigrationsService.run(config);
+    const newConfig = MigrationsService.run(config).migratedConfig;
```
Comment (viceice) [inline lib/config/migration.ts:36]: why you don't use the `isMigrated property?

File: lib/config/migrations/base/remove-property-migration.ts [hunk 2]
```diff
@@ -1,6 +1,18 @@
+import { RenovateConfig } from '../../types';
```
Comment (viceice) [inline lib/config/migrations/base/remove-property-migration.ts:1]:
```suggestion
import type { RenovateConfig } from '../../types';
```

File: lib/config/migrations/migrations-service.ts [hunk 3]
```diff
@@ -1,12 +1,14 @@
-import type { RenovateConfig } from '../types';
+import { dequal } from 'dequal';
+import type { MigratedConfig, RenovateConfig } from '../types';
+import { AbstractMigration } from './base/abstract-migration';
```
Comment (viceice) [inline lib/config/migrations/migrations-service.ts:3]:
```suggestion
import type { AbstractMigration } from './base/abstract-migration';
```

File: lib/config/migrations/types.ts [hunk 4]
```diff
@@ -1,4 +1,8 @@
-export interface Migration {
-  readonly propertyName: string;
-  run(): void;
+import { RenovateConfig } from './../types';
+import { AbstractMigration } from './base/abstract-migration';
```
Comment (viceice) [inline lib/config/migrations/types.ts:2]: I think leaving and using the `Migration` interface is better code.

File: lib/config/migration.ts (hunk 1 above)
Comment (pret-a-porter) [reply lib/config/migration.ts:36]: It is defined below in this file as `const isMigrated = !dequal(config, migratedConfig);`. For now `isMigrated` introduced inside `MigrationsService` for safely moving unit tests to separate files.

File: lib/config/migrations/types.ts [hunk 5]
```diff
@@ -1,4 +1,12 @@
+import { RenovateConfig } from './../types';
```
Comment (viceice) [inline lib/config/migrations/types.ts:1]:
```suggestion
import type { RenovateConfig } from './../types';
```

**Review by viceice — APPROVED** [review]

**Pull request merged**